File: python/FluidSolver.py
from __future__ import division, print_function
import os
import sys
import argparse
import configuration_file as config
from thetaScheme import perform_partitioned_implicit_trapezoidal_rule_step, perform_partitioned_implicit_euler_step
import numpy as np
import tubePlotting

import matplotlib
import matplotlib.pyplot as plt
import matplotlib.animation as manimation

from output import writeOutputToVTK

import precice
from precice import *

parser = argparse.ArgumentParser()
parser.add_argument("configurationFileName", help="Name of the xml precice configuration file.", nargs='?', type=str, default="precice-config.xml")
parser.add_argument("--write-vtk", help="Save vtk files of each timestep in the 'VTK' folder.", action='store_true')
parser.add_argument("--enable-plot", help="Show a continuously updated plot of the tube while simulating.", action='store_true')
parser.add_argument("--write-video", help="Save a video of the simulation as 'writer_test.mp4'. \
                    NOTE: This requires 'enable_plot' to be active!", action='store_true')

try:
    args = parser.parse_args()
except SystemExit:
    print("")
    print("Did you forget adding the precice configuration file as an argument?")
    print("Try '$ python FluidSolver.py precice-config.xml'")
    quit()

output_mode = config.OutputModes.VTK if args.write_vtk else config.OutputModes.OFF
plotting_mode = config.PlottingModes.VIDEO if args.enable_plot else config.PlottingModes.OFF
if args.write_video and not args.enable_plot:
    print("")
    print("To create a video it is required to enable plotting for this run.")
    print("Please supply both the '--enable-plot' and '--write-video' flags.")
    quit()
writeVideoToFile = True if args.write_video else False

print("Starting Fluid Solver...")

configFileName = args.configurationFileName

N = config.n_elem
dx = config.L / N  # element length

print("N: " + str(N))

solverName = "FLUID"

print("Configure preCICE...")
interface = precice.Interface(solverName, configFileName, 0, 1)
print("preCICE configured...")

dimensions = interface.get_dimensions()

velocity = config.velocity_in(0) * np.ones(N+1)
velocity_n = config.velocity_in(0) * np.ones(N+1)
pressure = config.p0 * np.ones(N+1)
pressure_n = config.p0 * np.ones(N+1)
crossSectionLength = config.a0 * np.ones(N+1)
crossSectionLength_n = config.a0 * np.ones(N+1)


if plotting_mode == config.PlottingModes.VIDEO:
    fig, ax = plt.subplots(1)
    if writeVideoToFile:
        FFMpegWriter = manimation.writers['imagemagick']
        metadata = dict(title='PulseTube')
        writer = FFMpegWriter(fps=15, metadata=metadata)
        writer.setup(fig, "writer_test.mp4", 100)

meshID = interface.get_mesh_id("Fluid_Nodes")
crossSectionLengthID = interface.get_data_id("CrossSectionLength", meshID)
pressureID = interface.get_data_id("Pressure", meshID)

vertexIDs = np.zeros(N+1)
grid = np.zeros([N+1, dimensions])

grid[:, 0] = np.linspace(0, config.L, N+1)  # x component
grid[:, 1] = 0  # y component, leave blank

vertexIDs = interface.set_mesh_vertices(meshID, grid)

t = 0

print("Fluid: init precice...")
precice_tau = interface.initialize()

if interface.is_action_required(action_write_initial_data()):
    interface.write_block_scalar_data(pressureID, vertexIDs, pressure)
    interface.fulfilled_action(action_write_initial_data())

interface.initialize_data()

if interface.is_read_data_available():
    interface.read_block_scalar_data(crossSectionLengthID, vertexIDs, crossSectionLength)

crossSectionLength_n = np.copy(crossSectionLength)
velocity_n = config.velocity_in(0) * crossSectionLength_n[0] * np.ones(N+1) / crossSectionLength_n  # initialize such that mass conservation is fulfilled

print(crossSectionLength_n)

while interface.is_coupling_ongoing():
    # When an implicit coupling scheme is used, checkpointing is required
    if interface.is_action_required(action_write_iteration_checkpoint()):
        interface.fulfilled_action(action_write_iteration_checkpoint())

    velocity, pressure, success = perform_partitioned_implicit_euler_step(velocity_n, pressure_n, crossSectionLength_n, crossSectionLength, dx, precice_tau, config.velocity_in(t + precice_tau), custom_coupling=False)
    interface.write_block_scalar_data(pressureID, vertexIDs, pressure)
    precice_tau = interface.advance(precice_tau)
    interface.read_block_scalar_data(crossSectionLengthID, vertexIDs, crossSectionLength)

    if interface.is_action_required(action_read_iteration_checkpoint()): # i.e. not yet converged
        interface.fulfilled_action(action_read_iteration_checkpoint())
    else:  # converged, timestep complete
        t += precice_tau
        if plotting_mode is config.PlottingModes.VIDEO:
            tubePlotting.doPlotting(ax, crossSectionLength_n, velocity_n, pressure_n, dx, t)
            if writeVideoToFile:
                writer.grab_frame()
            ax.cla()
        velocity_n = np.copy(velocity)
        pressure_n = np.copy(pressure)
        crossSectionLength_n = np.copy(crossSectionLength)
        if output_mode is config.OutputModes.VTK:
            writeOutputToVTK(t, "fluid", dx, N+1, datanames=["velocity", "pressure", "crossSection"], data=[velocity_n, pressure_n, crossSectionLength_n])

print("Exiting FluidSolver")

if plotting_mode is config.PlottingModes.VIDEO and writeVideoToFile:
    writer.finish()

interface.finalize()

File: cxx/FluidSolver_Serial/fluid_solver.cpp
#include "FluidSolver.h"
#include "fluid_nl.h"

#include "precice/SolverInterface.hpp"
#include <iostream>
#include <mpi.h>

using namespace precice;
using namespace precice::constants;

int main(int argc, char** argv)
{

  std::cout << "Starting Fluid Solver..." << std::endl;
  if (argc != 5 && argc != 6) {
    std::cout << std::endl;
    std::cout << "Fluid: Usage: mpiexec -np <#procs> " << argv[0] << " <configurationFileName> <N> <tau> <kappa> -parallel" << std::endl;
    std::cout << "or" << std::endl;
    std::cout << "Usage: " << argv[0] << " configurationFileName> <N> <tau> <kappa>" << std::endl;
    std::cout << std::endl;
    std::cout << "N:     Number of mesh elements, needs to be equal for fluid and structure solver." << std::endl;
    std::cout << "tau:   Dimensionless time step size." << std::endl;
    std::cout << "kappa: Dimensionless structural stiffness." << std::endl;

    return -1;
  }

  std::string configFileName(argv[1]);
  int domainSize = atoi(argv[2]); //N
  int chunkLength = domainSize + 1; //serial run
  double tau = atof(argv[3]);
  double kappa = atof(argv[4]);

  std::string solverName = "FLUID";

  std::string outputFilePrefix = "Postproc/out_fluid"; //extra

    
  int gridOffset, rank = 0, size = 1;


  if (argc == 6){
	  MPI_Init(&argc, &argv);
  	MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  	MPI_Comm_size(MPI_COMM_WORLD, &size);
    outputFilePrefix += std::to_string(rank); //extra

    if ((domainSize + 1) % size == 0) {
        chunkLength = (domainSize + 1) / size;
        gridOffset = rank * chunkLength;
    } else if (rank < (domainSize + 1) % size) {
        chunkLength = (domainSize + 1) / size + 1;
        gridOffset = rank * chunkLength;
      } else {
        chunkLength = (domainSize + 1) / size;
        gridOffset = ((domainSize + 1) % size) * ((domainSize + 1) / size + 1) + (rank - ((domainSize + 1) % size)) * (domainSize + 1) / size;
      }
  }



  SolverInterface interface(solverName, configFileName, rank , size);
  double *velocity, *velocity_n, *pressure, *pressure_n, *crossSectionLength, *crossSectionLength_n;

  velocity             = new double[chunkLength];
  velocity_n           = new double[chunkLength];
  pressure             = new double[chunkLength]; 
  pressure_n           = new double[chunkLength];
  crossSectionLength   = new double[chunkLength];
  crossSectionLength_n = new double[chunkLength];

  int *vertexIDs;
  vertexIDs = new int[chunkLength];

  int dimensions = interface.getDimensions();
  int meshID = interface.getMeshID("Fluid_Nodes");
  int pressureID = interface.getDataID("Pressure", meshID);
  int crossSectionLengthID = interface.getDataID("CrossSectionLength", meshID);

  
  double* grid;
  grid = new double[dimensions * chunkLength];
  
  for (int i = 0; i < chunkLength; i++) {
    pressure[i] = 0.0;
    pressure_n[i] = 0.0;
    crossSectionLength[i] = 1.0;
    crossSectionLength_n[i] = 1.0;
    velocity[i] = 1.0 / kappa;
    velocity_n[i] = 1.0 / kappa;
  }

  if (argc==6){
    for (int i = 0; i < chunkLength; i++) {
      for (int j = 0; j < dimensions; j++) {
          grid[i * dimensions + j] = j == 0 ? gridOffset + (double)i : 0.0;
      }
    
    }
  } else {
    for (int i = 0; i < chunkLength; i++) {
      for (int j = 0; j < dimensions; j++) {
          grid[i * dimensions + j] = i * (1 - j);
      }
    }
  }


  interface.setMeshVertices(meshID, chunkLength, grid, vertexIDs);

  std::cout << "Initialize preCICE..." << std::endl;
  interface.initialize();

  double t = 0.0;
  double dt = 0.01;

  if (interface.isActionRequired(actionWriteInitialData())) {
    interface.writeBlockScalarData(pressureID, chunkLength, vertexIDs, pressure);
    interface.markActionFulfilled(actionWriteInitialData());
  }

  interface.initializeData();

  if (interface.isReadDataAvailable()) {
    interface.readBlockScalarData(crossSectionLengthID, chunkLength, vertexIDs, crossSectionLength);
  }

  int out_counter = 0;

  while (interface.isCouplingOngoing()) {
    int convergenceCounter = 0;
    if (interface.isActionRequired(actionWriteIterationCheckpoint())) {
      interface.markActionFulfilled(actionWriteIterationCheckpoint());
    }

    if (argc == 6){
	    fluidComputeSolution(rank, size, domainSize, chunkLength, kappa, tau, 0.0, t+dt,
      pressure, pressure, pressure,
      crossSectionLength, crossSectionLength,
      velocity, velocity);
    } else {
      fluid_nl(crossSectionLength, crossSectionLength,  
	    velocity, velocity,                      
	    pressure, pressure,            
	    t, domainSize, kappa, tau); 
    }
    

    interface.writeBlockScalarData(pressureID, chunkLength, vertexIDs, pressure);

    interface.advance(dt);

    interface.readBlockScalarData(crossSectionLengthID, chunkLength, vertexIDs, crossSectionLength);

    if (interface.isActionRequired(actionReadIterationCheckpoint())) { // i.e. not yet converged
      interface.markActionFulfilled(actionReadIterationCheckpoint());
      convergenceCounter++;
    } else {
      t += dt;
      for (int i = 0; i < chunkLength; i++) {
        pressure_n[i] = pressure[i];
        velocity_n[i] = velocity[i];
        crossSectionLength_n[i] = crossSectionLength[i];
      }
      //write_vtk(t, out_counter, outputFilePrefix.c_str(), domainSize, grid, velocity_n, pressure_n, crossSectionLength_n);
      write_vtk(t, out_counter, outputFilePrefix.c_str(), chunkLength, grid, velocity_n, pressure_n, crossSectionLength_n);

      out_counter++;
    }
  }

  delete [] velocity;
  delete [] velocity_n;
  delete [] pressure;
  delete [] pressure_n;
  delete [] crossSectionLength;
  delete [] crossSectionLength_n;
  delete [] vertexIDs;
  delete [] grid;
  interface.finalize();
  if (argc == 6){
    MPI_Finalize();

  }

  return 0;
}